src/jit/decompose.cpp

```diff
@@ -0,0 +1,1189 @@
+XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+XX                                                                           XX
+XX                               Decompose                                   XX
+XX                                                                           XX
+XX Code to decompose 64-bit LONG operations on 32-bit platforms into         XX
+XX multiple single-register operations so individual register usage and      XX
+XX requirements are explicit for LSRA.                                       XX
+XX                                                                           XX
+XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+*/
+
+#include "jitpch.h"
+#ifdef _MSC_VER
+#pragma hdrstop
+#endif
+
+#ifndef LEGACY_BACKEND // This file is ONLY used for the RyuJIT backend that uses the linear scan register allocator
+#ifndef _TARGET_64BIT_ // Decompose is only used on 32-bit platforms
+
+#include "decompose.h"
+
+//------------------------------------------------------------------------
+
+//------------------------------------------------------------------------
+// PrepareForDecomposition: Do one-time preparation required for LONG decomposition.
+// Namely, promote long variables to multi-register structs.
+//
+// Arguments:
+//    None
+//
+// Return Value:
+//    None.
+//
+void Decompose::PrepareForDecomposition()
+{
+    comp->lvaPromoteLongVars();
+}
+
+
+//------------------------------------------------------------------------
+// DecomposeBlock: Do LONG decomposition to all the statements in the given block.
+// This must be done before lowering the block, as decomposition can insert
+// additional statements.
+//
+// Decomposition is done as a post-order tree walk. Lower levels of the tree can
+// create new nodes that need to be further decomposed at higher levels. That is,
+// the decomposition "bubbles up" the tree.
+//
+// Arguments:
+//    block - the block to process
+//
+// Return Value:
+//    None.
+//
+void Decompose::DecomposeBlock(BasicBlock* block)
+{
+    for (GenTree* stmt = block->bbTreeList; stmt != nullptr; stmt = stmt->gtNext)
+    {
+        if (stmt->gtFlags & GTF_STMT_SKIP_LOWER)
+        {
+            continue;
+        }
+
+#ifdef DEBUG
+        if (comp->verbose)
+        {
+            printf("Decomposing BB%02u, stmt id %u\n", block->bbNum, stmt->gtTreeID);
+        }
+#endif // DEBUG
+
+        DecomposeStmt(stmt->AsStmt());
+    }
+}
+
+
+//------------------------------------------------------------------------
+// DecomposeStmt: Do LONG decomposition to a statement tree.
+//
+// Arguments:
+//    stmt - the statement to process
+//
+// Return Value:
+//    None.
+//
+void Decompose::DecomposeStmt(GenTreeStmt* stmt)
+{
+    comp->compCurStmt = stmt;   // publish the current statement globally (is this really necessary?)
```

Comment:
> is this really necessary?

Good question. Seems like it isn't used anywhere in the decomposition code. Still, it might be used by some `Compiler` function the decomposition code calls.